File: levenshtein.go
// Levenshtein distance in Go

// i'm just going to plop this func into the main file for now... being lazy!

package editdist

func EditDistance(a, b string) int {
	switch {
		case len(a) == 0 {
			return len(b)
		}
		case len(b) == 0 {
			return len(a)
		}
		default {
			var i, j, k int
			// setting i
			if a[0] == b[0] {
				i = Levenshtein(a[1:], b[1:])
			} else {
				i = 1 + Levenshtein(a[1:], b[1:])
			}
			// setting j
			j = 1 + Levenshtein(a[1:], b)
			// setting k
			j = 1 + Levenshtein(a, b[1:])
			
			// return the min
			min := 0
			if i < j {
				min  = i
			} else {
				min = j
			}
			if k < min {
				return k
			} else {
				return min
			}
		}
	}
}

File: levenshtein.rb

=begin info

  module Levenshtein
  calculates Levenshtein distance between 2 strings or arrays
  
  mostly copied from:
  http://en.wikibooks.org/wiki/Algorithm_implementation/Strings/Levenshtein_distance#Ruby
  though it's been refactored to make it more object-like

=end

module Levenshtein
	def levenshtein other
	  case
	    when self.empty? then other.length
	    when other.empty? then self.length
	    else [(self[0] == other[0] ? 0 : 1) + self[1..-1].levenshtein(other[1..-1]),
	          1 + self[1..-1].levenshtein(other),
	          1 + self.levenshtein(other[1..-1])].min
	  end
	end
	alias :edit_distance :levenshtein
  alias :edit_dist :levenshtein
end

# including the module in string, to make life easier...
String.class_eval do
  include Levenshtein
end

# and even though i don't need it here i'll include it in array also, cause i can!
Array.class_eval do
  include Levenshtein
end
